### ISSUE TITLE
Resource attributes accessed by '[each.key]' are incorrectly referenced when 'try' is used

### ISSUE THREAD

**Issue by bennyguk** [hashicorp/terraform#32800]
### Terraform Version

```shell
Terraform v1.3.9
on windows_amd64
+ provider registry.terraform.io/hashicorp/null v3.2.1
```

### Terraform Configuration Files

```terraform {
  required_providers {
    null = {
      source  = "hashicorp/null"
      version = ">= 3.0"
    }
  }
  required_version = "= 1.3.9"
}

variable "subnets" {
  default = [
    {
      name = "subnet1",
      cidr = "10.68.57.0/24",
      type = "private"
    },
    {
      name = "subnet2",
      cidr = "10.68.56.128/27",
      type = "private"
    },
    {
      name = "subnet3",
      cidr = "10.68.56.160/27",
      type = "private"
    },
    {
      name = "subnet4",
      cidr = "10.68.58.0/24",
      type = "private"
    },
    {
      name            = "appgw",
      cidr            = "10.68.59.0/24"
      type            = "none"
      is_appgw_subnet = true
    }
  ]
}

resource "null_resource" "subnet" {
  for_each = local.require_nsg_creation
  triggers = {
    name = each.value.name
  }
}

resource "null_resource" "nsg_rule" {
  for_each = { for subnet in var.subnets : subnet.name => subnet if try(subnet.is_appgw_subnet, false) }
  triggers = {
    network_security_group_name = try(each.value.assigned_nsg, null_resource.subnet[each.key].triggers.name)
  }
}

locals {
  require_nsg_creation = { for subnet in var.subnets : subnet.name => subnet if alltrue([try(subnet.assigned_nsg, "") == "", try(subnet.require_nsg_binding, true), !contains(["AzureBastionSubnet", "AzureFirewallSubnet", "GatewaySubnet"], subnet.name)]) }
}
```

### Debug Output

https://gist.github.com/bennyguk/3cd36a20e29361241de1609843dc5379

### Expected Behavior

When adding another item to the `subnets` variable with `is_appgw_subnet = true`, e.g:
```
    {
      name            = "appgw2",
      cidr            = "10.68.60.0/24"
      type            = "none"
      is_appgw_subnet = true
    }
```
`null_resource.nsg_rule["appgw2"]` and `null_resource.subnet["appgw2"]` are created without affecting the pre-existing `null_resource.nsg_rule["appgw"]`:

```
Terraform will perform the following actions:

# null_resource.nsg_rule["appgw2"] will be created
  + resource "null_resource" "nsg_rule" {
      + id       = (known after apply)
      + triggers = {
          + "network_security_group_name" = (known after apply)
        }
    }

  # null_resource.subnet["appgw2"] will be created
  + resource "null_resource" "subnet" {
      + id       = (known after apply)
      + triggers = {
          + "name" = "appgw2"
        }
    }
```

### Actual Behavior

Terraform wants to replace `null_resource.nsg_rule["appgw"]` because the `network_security_group_name` attribute is changing:

```
Terraform will perform the following actions:

  # null_resource.nsg_rule["appgw"] must be replaced
-/+ resource "null_resource" "nsg_rule" {
      ~ id       = "5359056016556765851" -> (known after apply)
      ~ triggers = { # forces replacement
          ~ "network_security_group_name" = "appgw" -> (known after apply)
        }
    }

  # null_resource.nsg_rule["appgw2"] will be created
  + resource "null_resource" "nsg_rule" {
      + id       = (known after apply)
      + triggers = {
          + "network_security_group_name" = (known after apply)
        }
    }

  # null_resource.subnet["appgw2"] will be created
  + resource "null_resource" "subnet" {
      + id       = (known after apply)
      + triggers = {
          + "name" = "appgw2"
        }
    }
```
Looking at the debug log, it seems that Terraform references "appgw2" in the `subnets` variable even though `network_security_group_name` attribute should reference the same name each time using `null_resource.subnet[each.key].triggers.name` and in this case, it should remain to be "appgw":

`[DEBUG] ReferenceTransformer: "null_resource.nsg_rule[\"appgw\"]" references: [var.subnets null_resource.subnet (expand) null_resource.subnet["appgw2"]]`

When removing `try` e.g. `network_security_group_name = null_resource.subnet[each.key].triggers.name` the plan works as expected and `null_resource.nsg_rule["appgw"]` is no longer modified.

### Steps to Reproduce

1. terraform init
2. terraform apply
3. add another item to the `subnets` variable with the `is_appgw_subnet` attribute set to `true` such as:
```
    {
      name            = "appgw2",
      cidr            = "10.68.60.0/24"
      type            = "none"
      is_appgw_subnet = true
    }
```
4. terraform plan

**Comment by jbardin** [comment]
Hi @bennyguk,

Thanks for filing the issue, and the complete example! What you are seeing here is result of the special evaluation required within `can` and `try` functions, so I'm going to combine this with the existing issue #31646.

In the meantime, you can avoid the use of `try` in these situations by defining the expected types rather than guessing at runtime. For example, if the `subnets` variable has a type definition of:

```
  type = list(object({
    name = string
    cidr = string
    type = string
    assigned_nsg = optional(string),
    is_appgw_subnet = optional(bool, false),
  }))
```

you be certain that the object attributes always exist eliminating the `try` calls. 

Thanks!

**Comment by crw** [comment]
Closed as duplicate of #31646.
